File: NthNodeFromEnd.py
# Simple Python3 program to find 
# n'th node from end 
class Node: 
	def __init__(self, new_data): 
		self.data = new_data 
		self.next = None
	
class LinkedList: 
	def __init__(self): 
		self.head = None

	# createNode and and make linked list 
	def push(self, new_data): 
		new_node = Node(new_data) 
		new_node.next = self.head 
		self.head = new_node 

	# Function to get the nth node from 
	# the last of a linked list 
	def printNthFromLast(self, n): 
		temp = self.head # used temp variable 
		
		length = 0
		while temp is not None: 
			temp = temp.next
			length += 1
		
		# print count 
		if n > length: # if entered location is greater 
					# than length of linked list 
			print('Location is greater than the' +
						' length of LinkedList') 
			return
		temp = self.head 
		for i in range(0, length - n): 
			temp = temp.next
		print(temp.data) 

# Driver Code		 
llist = LinkedList() 
llist.push(20) 
llist.push(4) 
llist.push(15) 
llist.push(35) 
llist.printNthFromLast(4) 

File: NthNodeFromEnd.java
// Simple Java program to find n'th node from end of linked list 
class LinkedList { 
	Node head; // head of the list 

	/* Linked List node */
	class Node { 
		int data; 
		Node next; 
		Node(int d) 
		{ 
			data = d; 
			next = null; 
		} 
	} 

	/* Function to get the nth node from the last of a 
	linked list */
	void printNthFromLast(int n) 
	{ 
		int len = 0; 
		Node temp = head; 

		// 1) count the number of nodes in Linked List 
		while (temp != null) { 
			temp = temp.next; 
			len++; 
		} 

		// check if value of n is not more than length of 
		// the linked list 
		if (len < n) 
			return; 

		temp = head; 

		// 2) get the (len-n+1)th node from the beginning 
		for (int i = 1; i < len - n + 1; i++) 
			temp = temp.next; 

		System.out.println(temp.data); 
	} 

	/* Inserts a new Node at front of the list. */
	public void push(int new_data) 
	{ 
		/* 1 & 2: Allocate the Node & 
				Put in the data*/
		Node new_node = new Node(new_data); 

		/* 3. Make next of new Node as head */
		new_node.next = head; 

		/* 4. Move the head to point to new Node */
		head = new_node; 
	} 

	/*Drier program to test above methods */
	public static void main(String[] args) 
	{ 
		LinkedList llist = new LinkedList(); 
		llist.push(20); 
		llist.push(4); 
		llist.push(15); 
		llist.push(35); 

		llist.printNthFromLast(4); 
	} 
} 
